File: tools/visualvm-web/src/main/java/org/toolkit/visualvm/web/controller/CPUController.java
/* (C)2023*/
package org.toolkit.visualvm.web.controller;

import io.javalin.http.sse.SseClient;
import jakarta.inject.Inject;
import java.util.Queue;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentLinkedQueue;
import lombok.SneakyThrows;
import org.toolkit.visualvm.web.service.ICPUService;
import oshi.SystemInfo;
import oshi.hardware.CentralProcessor;

public class CPUController {

	@Inject
	private ICPUService cpuService;

	@Inject
	private SystemInfo systemInfo;

	private final Queue<SseClient> clients = new ConcurrentLinkedQueue<SseClient>();

	@SneakyThrows
	public void cpuLoadSse(SseClient sseClient) {
		clients.add(sseClient);
		sseClient.keepAlive();
		sseClient.onClose(() -> clients.remove(sseClient));
		CompletableFuture.runAsync(() -> {
			CentralProcessor processor = systemInfo.getHardware().getProcessor();

			// 获取上一次 CPU 时间戳
			long[] prevTicks = processor.getSystemCpuLoadTicks();
			// 循环获取 CPU 占用
			while (true) {
				// 获取当前 CPU 时间戳
				long[] ticks = processor.getSystemCpuLoadTicks();

				// 计算 CPU 占用百分比
				double cpuUsage = processor.getSystemCpuLoadBetweenTicks(prevTicks) * 100.0;
				sseClient.sendEvent("connected", cpuUsage);
				// 输出 CPU 占用百分比
				System.out.printf("Current CPU Usage: %.2f%%\n", cpuUsage);

				// 更新上一次 CPU 时间戳
				prevTicks = ticks;

				// 等待一段时间，可以根据需求调整
				try {
					Thread.sleep(1000);
				} catch (InterruptedException e) {
					throw new RuntimeException(e);
				}
			}
		});
	}
}

File: tools/visualvm-web/src/main/kotlin/org/nectar/visualvm/web/CpuController.kt
@file:OptIn(DelicateCoroutinesApi::class)

package org.nectar.visualvm.web

import com.fasterxml.jackson.databind.ObjectMapper
import io.javalin.http.sse.SseClient
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.slf4j.LoggerFactory
import oshi.SystemInfo
import java.lang.invoke.MethodHandles
import java.text.DecimalFormat
import java.util.concurrent.ConcurrentLinkedQueue

val clients = ConcurrentLinkedQueue<SseClient>()
private val logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass())
private val systemInfo = inject<SystemInfo>()
private val objectMapper = inject<ObjectMapper>()

fun sseBroadcast() {
    val processor = systemInfo.value.hardware.processor
    // 获取上一次 CPU 时间戳
    var prevTicks = processor.systemCpuLoadTicks
    val mapper = objectMapper.value
    // 使用协程
    GlobalScope.launch {
        while (true) {
            if (clients.isEmpty()) {
                continue
            }

            logger.atInfo().log("send:{}", clients)

            // 获取当前 CPU 时间戳
            val ticks = processor.systemCpuLoadTicks
            val cpuUsage = processor.getSystemCpuLoadBetweenTicks(prevTicks) * 100.0
            val decimalFormat = DecimalFormat("#.##")
            val formattedCpuUsage = decimalFormat.format(cpuUsage)
            // 更新上一次 CPU 时间戳
            prevTicks = ticks
            val a =
                object {
                    val key = "cpu"
                    val load = formattedCpuUsage
                }
            val load = mapper.writeValueAsString(a)
            // 发送事件给所有客户端
            clients.forEach {
                it.sendEvent(load)
            }

            delay(1000) // 延迟一秒后再次循环
        }
    }
}

fun cpuReadTimeLoad(client: SseClient) {
    clients.add(client)
    client.keepAlive()
    client.onClose { clients.remove(client) }
}
